spm: Add debug option to dump non-secure logs on secure fault

nordic-krch opened the pull request:
When deferred logging is enabled and non-secure application triggers a secure fault then application reboots silently without any debugging information. That is because:
- SPM can no longer print to UARTE0 because it was reconfigured to non-secure and locked.
- Non-secure app contains unprocessed logs in the buffer but it is never notified and has no chance to dump logs.

It makes it very hard to debug a faulty scenario. PR fixes that in 2 steps:
- enables runtime UARTE0 permission reconfiguration allowing to print to the console from secure fault handler. It is done by adding a minimal uart driver which wraps poll_out with securing UART peripheral and restoring previous state after polling out.
- calling in secure fault handler a function provided by the non-secure application. Function calls `LOG_PANIC`.

As this feature is considered non-secure, it is optional and by default disabled (`CONFIG_SPM_NS_DEBUG_LOGS_DUMP`).

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-3933
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

jhn-nordic commented:
@nordic-krch Could you add overlay for the thingy91_nrf9160 as well?

hakonfam commented:
Does this fix NCSIDB-114 which is also in the known-issues?

nordic-krch commented:
@hakonfam no, if spm would use logging mode which fetches timestamp it will still crash but spm is now using minimal logging without timestamping. That seems to be enough for time being.

hakonfam commented on subsys/spm/Kconfig:
```diff
@@ -55,6 +55,14 @@ config SPM_BOOT_SILENTLY
 
 if IS_SPM || SPM
+
+config SPM_NS_DEBUG_LOGS_DUMP
```
IMO there should be a warning from CMake if this is set. We might want to have a warning in kconfig as well. This is typically done for all options which must be disabled for production (debug keys, verbose assert etc).

hakonfam replied:
```
if SPM_NS_DEBUG_LOGS_DUMP:
    comment "SPM_NS_DEBUG_LOGS_DUMP can only be used for debugging"
```

hakonfam commented on subsys/spm/CMakeLists.txt:
```diff
@@ -4,6 +4,11 @@
 #
 
+if (CONFIG_SPM)
+zephyr_sources_ifdef(CONFIG_SPM_NS_DEBUG_LOGS_DUMP spm_ns.c)
```
This file has wrong indentation now.

hakonfam commented on subsys/spm/Kconfig:
```diff
@@ -55,6 +55,14 @@ config SPM_BOOT_SILENTLY
+	help
+	  When enabled secure fault handler is calling non-secure function to
+	  synchronously dump all pending logs. Feature is only for debugging
+	  purposes and is considered as security violation.
```
`considered a security`

hakonfam commented on include/spm.h:
```diff
@@ -41,6 +48,25 @@ void spm_jump(void);
  */
 void spm_config(void);
 
+/** @brief Set handler which is called on the fault handler on the secure parition.
```
What do you think about `which is called by the SPM fault handler` ? I don't we call the SPM `the secure partition`.

hakonfam commented on include/spm.h:
```diff
@@ -41,6 +48,25 @@ void spm_jump(void);
+ * @retval -ENOTSUP if feature is disabled.
+ * @retval 0 on success.
+ */
+int spm_ns_set_fatal_error_handler(spm_ns_on_fatal_error_t handler);
```
Do you think that `spm_set_ns_fatal_error_handler` is a better name? Currently it sounds like you are setting the error handler of the non-secure spm.

hakonfam commented on lib/fatal_error/fatal_error.c:
```diff
@@ -5,6 +5,7 @@
  */
 
 #include <arch/cpu.h>
+#include <spm.h>
```
Should this be conditional to emphasize that the functionality is only used for certain configurations?

nordic-krch replied:
Personally, i don't like ifdefing headers, which leads to ifdefing code.

hakonfam commented on lib/fatal_error/fatal_error.c:
```diff
@@ -21,6 +22,10 @@ void k_sys_fatal_error_handler(unsigned int reason,
 
 	LOG_PANIC();
 
+	if (IS_ENABLED(CONFIG_IS_SPM) && IS_ENABLED(CONFIG_SPM_NS_DEBUG_LOGS_DUMP)) {
```
I don't want to start a war here, but our .clang-format is still 80 chars and AFAIU it is preferred to stay within 80 unless it makes the code less readable. IMO we should break this line to fit within 80, but feel free to ignore.

hakonfam commented on subsys/spm/spm.c:
```diff
@@ -252,7 +254,7 @@ static bool usel_or_split(uint8_t id)
 }
 
-static int spm_config_peripheral(uint8_t id, bool dma_present)
+static int spm_config_peripheral(uint8_t id, bool dma_present, bool lock)
```
I don't like this change, since it opens up for customers accidentally not locking a peripheral. AFAICS the only reason not to lock a configuration is due to debugging like we do in this PR. I would prefer if we add a separate function, which is only compiled in when certain configurations are set (such as the one in this PR). This way it is clear that not locking your peripheral is something that is ONLY done in development.

hakonfam replied:
Alternatively we could have a separate function for configuring the UART as it can be considered a special case. I just don't want to expose the option of not locking the configuration here.

nordic-krch replied:
> it opens up for customers accidentally not locking a peripheral. 

It's not my area so i will end up doing what security guys are telling but my opinion first:
If a customer modifies SPM he should know what is doing. Parameter name is quite obvious and peripheral ownership can only be changed from secure state so no secure hole here as well.

nordic-krch replied:
reverted that. Added dedicated `spm_config_unlocked_peripheral()` which is used for uarte0 only.

hakonfam commented on subsys/spm/spm.c:
```diff
@@ -567,3 +568,60 @@ void spm_config(void)
+
+#ifdef CONFIG_UART_0_NRF_UARTE
+
+static bool spm_periph_reconf(uint8_t id, bool secure)
```
It is not clear what this function does, can you provide a @brief with a oneliner explaining it?

hakonfam commented on subsys/spm/spm.c:
```diff
@@ -567,3 +568,60 @@ void spm_config(void)
+{
+	bool sec = (NRF_SPU->PERIPHID[id].PERM & PERIPH_SEC);
+
+	NRF_SPU->PERIPHID[id].PERM = BIT(5) | (secure ? PERIPH_SEC : 0);
```
What is `BIT(5)` should we have a define for it?

hakonfam commented on subsys/spm/spm.c:
```diff
@@ -567,3 +568,60 @@ void spm_config(void)
+
+static bool spm_periph_reconf(uint8_t id, bool secure)
+{
+	bool sec = (NRF_SPU->PERIPHID[id].PERM & PERIPH_SEC);
```
What about `previous_sec_configuration = ...`  ?

`sec` is a bit undescriptive.

hakonfam commented on subsys/spm/spm.c:
```diff
@@ -567,3 +568,60 @@ void spm_config(void)
+	NRF_UARTE_Type *regs;
+};
+
+#ifdef CONFIG_UART_0_NRF_UARTE
```
Shouldn't all of this be conditionally included only if you need to dump the non secure logs?

hakonfam commented on samples/spm/boards/nrf5340dk_nrf5340_cpuapp.overlay:
```diff
@@ -0,0 +1,17 @@
+&uart0 {
+	status = "okay";
+
+	spm_uart: nrf-spm-uart {
```
Should we have a comment here explaining why this is needed? Also, is this needed if non secure logs should not be dumped?

oyvindronningstad commented on subsys/spm/spm_ns.c:
```diff
@@ -0,0 +1,30 @@
+/*
```
This file could instead be placed in subsys/nonsecure, which already has nonsecure parts of SPM functionality.

oyvindronningstad commented on subsys/spm/spm_ns.c:
```diff
@@ -0,0 +1,30 @@
+static int spm_ns_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	int err = spm_ns_set_fatal_error_handler(log_dump);
```
Where is the implementation of spm_ns_set_fatal_error_handler()? I can't find it in the PR, sorry if I'm missing something :P.

nordic-krch replied:
that is a mistery :) somehow it got lost. Fixed now.

nordic-krch commented:
@oyvindronningstad @hakonfam PR updated. I've reworked it a bit and splitted into 2 fearures:
- shared UARTE0 (`CONFIG_SPM_SHARE_CONSOLE_UART` option in spm) which allows printing from spm even after jumping to non-secure app. Enabling this feature only results in secure fault report being printed.
- dumping ns logs from spm fatal error handler (`CONFIG_SPM_NS_DEBUG_LOGS_DUMP` must be set in spm and app). Requires `CONFIG_SPM_SHARE_CONSOLE_UART`.

By default features are disabled but i've added `spm_uart` to spm overlays, together with commented choosing of spm_uart as zephyr console. It must be uncommented when `CONFIG_SPM_SHARE_CONSOLE_UART` is enabled. There are build time asserts in the code if dt configuration is wrong.

hakonfam commented on samples/spm/boards/nrf5340pdk_nrf5340_cpuapp.overlay:
```diff
@@ -0,0 +1,24 @@
+/* SPDX-License-Identifier: LicenseRef-Nordic-5-Clause */
```
I think this can be removed now since PDK support was removed just prior to 1.5.0

nordic-krch commented on subsys/spm/Kconfig:
```diff
@@ -52,9 +52,19 @@ config PM_PARTITION_SIZE_SPM_SRAM
+	  When enabled, UARTE0 can be used for printing from secure domain
+	  after jumping to the non-secure application, e.g. from exception
+	  handler. If enabled, zephyr console must be changed do spm_uart in
+	  the device tree configuration.
```
it states `If enabled, zephyr console must be changed do spm_uart in the device tree configuration.` I didn't want to explicitly mention overlay here because overlay is in the sample and here we are in subsys.

hakonfam replied:
understood.

hakonfam commented on subsys/spm/Kconfig:
```diff
@@ -66,6 +76,21 @@ config SPM_SECURE_SERVICES
 	  of the SPM. This ensures that the values are in sync between the two
 	  images.
+
+config SPM_NS_DEBUG_LOGS_DUMP
```
Since you need to enable multiple configurations in the child image SPM when the parent sets this option, you should consider doing it automatically. 

Look at how the overlay file is added here https://github.com/nrfconnect/sdk-nrf/blob/master/samples/CMakeLists.txt#L60

then you can add a check for this option and add the overlay here https://github.com/nrfconnect/sdk-nrf/blob/master/samples/CMakeLists.txt#L21

With this in place a user simply needs to enable option this in their application to have the correct kconfig in the SPM child image.

One missing piece is the manual modification of the DT overlay, but if you improve the error code printed when the wrong console is selected that would be improved as well.

nordic-krch replied:
done, Added `samples/spm/spm_ns_debug_logs_dump.conf` which is added to spm if `SPM_NS_DEBUG_LOGS_DUMP` is enabled in the app. Also rephrased a bit build assert message.

hakonfam commented:
Please look here for a proposal on how to achieve this without manual intervention:

https://github.com/nordic-krch/fw-nrfconnect-nrf/pull/1

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -20,17 +20,29 @@ function(get_board_without_ns_suffix board_in board_out)
+# This can be used by a parent image to modify the variables passed to its
+# child images. This must be invoked before 'add_child_image(image)'
+function(child_image_append_conf image val type)
+  set(old_conf ${${image}_${type}})
+  string(FIND "${old_conf}" "${val}" found)
+  if (${found} EQUAL -1)
```
Will this work as expected ?
Imagine:
```
  child_image_append_conf(spm some_file.overlay DTC_OVERLAY_FILE)
```

then if user provides a custom setting, like:
`cmake .... -Dspm_DTC_OVERLAY_CONFIG="file1.overlay" <sample>`

then `old_conf` becomes: `file1.overlay` which will give `-1` when searching for this sting inside: `${val}` (== `some_file.overlay`) thus:
```
spm_OVERLAY_CONFIG="file1.overlay some_file.overlay"
```
now, if CMake is reinvoked, then `spm_OVERLAY_CONFIG` is now: `file1.overlay some_file.overlay`, but that string is nowhere to be found inside `${val}` (== `some_file.overlay`), thus after second cmake, we'll have:
```
spm_OVERLAY_CONFIG="file1.overlay some_file.overlay some_file.overlay"
```

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -20,17 +20,29 @@ function(get_board_without_ns_suffix board_in board_out)
+# Append a value to a variable prefixed with a child image name.
+# This can be used by a parent image to modify the variables passed to its
+# child images. This must be invoked before 'add_child_image(image)'
+function(child_image_append_conf image val type)
```
I'm not sure we should document this function.

I don't like such generic function, as people might start to use it for unintended purposes, and thus makes it even harder for our users to understand when / why / where settings are being defined.

We should use the common `boards/` and `child_image` config options as much as possible, and extend those if need be.

I know there is a need for the current `add_overlay_config` function and maybe even a `devicetree` overlay also if there really is a need there. (preference would still be a extension of: http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_multi_image.html#image-specific-variables to support devicetree overlay if we have a need)

If keeping this helper function, then clearly mark it `Internal use only`.

tejlmand reviewed (changes requested):
Discussed another approach offline.

Using `UART_CONSOLE_ON_DEV_NAME` which can be depending on `SPM_SHARE_CONSOLE_UART` setting will remove the need for `zephyr,console = &spm_uart` in the overlay.

This allows to always have the overlay present, and no need for making the overlay depending on Kconfig settings.

nordic-krch commented:
@tejlmand @hakonfam cmake extension dropped, added setting of uart console dev. Seems that it is the simplest to use solution. When user enables debug feature in non-secure app, spm is configured to use switchable uart and debug option are enabled there.

nordic-krch commented:
@tejlmand @oyvindronningstad can you take another look?

nordic-krch commented on lib/fatal_error/fatal_error.c:
```diff
@@ -21,6 +22,11 @@ void k_sys_fatal_error_handler(unsigned int reason,
 
+	if (IS_ENABLED(CONFIG_IS_SPM) &&
+	    IS_ENABLED(CONFIG_SPM_NS_DEBUG_LOGS_DUMP)) {
+		spm_ns_fatal_error_handler();
```
I don't see good alternative  because otherwise function from `zephyr/kernel/fatal.c`is called. It's also the default in ncs.

oyvindronningstad replied:
You can make sure the file is built `if (CONFIG_RESET_ON_FATAL_ERROR or CONFIG_SPM_NS_DEBUG_LOGS_DUMP)`, and encase the `sys_arch_reboot()` call in an ifdef.

oyvindronningstad commented on subsys/spm/Kconfig:
```diff
@@ -66,6 +76,21 @@ config SPM_SECURE_SERVICES
+	  When enabled secure fault handler is calling non-secure function to
+	  synchronously dump all pending logs. It must be enabled in spm and
+	  in non-secure application. Feature is only for debugging purposes and
+	  is considered a security violation. SPM_SHARE_CONSOLE_UART must be
+	  enabled in spm.
```
> SPM_SHARE_CONSOLE_UART must be enabled in spm.

From what I can tell, SPM_SHARE_CONSOLE_UART isn't really needed. To me it seems like SPM_SHARE_CONSOLE_UART and SPM_NS_DEBUG_LOGS_DUMP are completely independent.

nordic-krch replied:
when UARTE is set to non-secure and locked before jumping to ns then when secure fault handler occurs it jumps to spm and attempts to access non-secure UARTE providing secure buffer pointer. It seems that it generates another secure fault since when `SPM_SHARE_CONSOLE_UART` is disabled i see only reset and no data printed.

oyvindronningstad replied:
SPM_NS_DEBUG_LOGS_DUMP is only for dumping NS logs right?

oyvindronningstad commented on subsys/spm/Kconfig:
```diff
@@ -66,6 +76,21 @@ config SPM_SECURE_SERVICES
 	  of the SPM. This ensures that the values are in sync between the two
 	  images.
+
+config SPM_NS_DEBUG_LOGS_DUMP
```
Did you consider presenting this as two different features?
1: In SPM: A general fatal error handler calling into NS when an S error happens
2: In NS: Using the SPM feature to dump logs.
The function naming on the SPM side is already quite generic, so it seems natural.
Also, I found it a bit confusing that the kconfig is used in both SPM and NS, and it does different things in those two.

nordic-krch replied:
I applied this proposal.

oyvindronningstad commented on subsys/spm/spm.c:
```diff
@@ -567,3 +582,93 @@ void spm_config(void)
+static spm_ns_on_fatal_error_t ns_on_fatal_handler;
+
+__TZ_NONSECURE_ENTRY_FUNC
+int spm_set_ns_fatal_error_handler(spm_ns_on_fatal_error_t handler)
```
I'm thinking maybe this should go into secure_services.c.
I think this can fail if SECURE_SERVICES is turned off, since CONFIG_ARM_FIRMWARE_USES_SECURE_ENTRY_FUNCS (in NS app) and CONFIG_ARM_FIRMWARE_HAS_SECURE_ENTRY_FUNCS (in SPM) need to be set for nonsecure entry functions to work.
If you make this a secure service, you don't have to worry about that.

nordic-krch replied:
moved to secure services.

oyvindronningstad commented on subsys/spm/spm.c:
```diff
@@ -567,3 +582,93 @@ void spm_config(void)
+}
+#endif /* CONFIG_SPM_NS_DEBUG_LOGS_DUMP */
+
+#if CONFIG_SPM_SHARE_CONSOLE_UART
```
This section might merit being placed in its own separate file. What do you think?

nordic-krch commented:
@oyvindronningstad i applied your comments. Can you take another look?

nordic-krch commented on lib/fatal_error/fatal_error.c:
```diff
@@ -21,8 +26,19 @@ void k_sys_fatal_error_handler(unsigned int reason,
+	} else {
+		for (;;) {
+			/* Spin endlessly */
+		}
```
done, except `LOG_PANIC` which is already called.

nordic-krch commented:
@tejlmand can you finish review?

tejlmand commented on samples/spm/prj.conf:
```diff
@@ -16,3 +16,8 @@ CONFIG_MAIN_STACK_SIZE=4096
+# When enabled, spm can print to UARTE0 even after switching to non-secure, e.g.
+# from secure fault handler. zephyr,console chosen must be uncommented in the
+# overlay file.
+#CONFIG_SPM_SHARE_CONSOLE_UART=y
```
wouldn't it be better to place this in a dedicated Kconfig fragment, so that users don't have to modify `prj.conf` to enable this, but can enable this through the use of `OVERLAY_CONFIG` ?

That is, make a `uart_share.config.conf`:
```
# This Kconfig fragment allows spm to print to UARTE0 even after switching to non-secure, e.g.
# from secure fault handler.
# To enable when building spm: `west build ... -- -DOVERLAY_CONFIG=uart_share.config.conf
# To enable when building spm in multiimage context: `west build ... -- -spm_DOVERLAY_CONFIG=uart_share.config.conf
CONFIG_SPM_SHARE_CONSOLE_UART=y
```

tejlmand commented on samples/spm/prj.conf:
```diff
@@ -16,3 +16,8 @@ CONFIG_MAIN_STACK_SIZE=4096
+
+# When enabled, spm can print to UARTE0 even after switching to non-secure, e.g.
+# from secure fault handler. zephyr,console chosen must be uncommented in the
+# overlay file.
```
> zephyr,console chosen must be uncommented in the overlay file.

Should this has been removed ?

nordic-krch replied:
it was already removed.

tejlmand reviewed (changes requested):
Minor proposals.

tejlmand commented:
Will there be a followup PR which documents this feature ?

tejlmand reviewed (approved):
Code approved, but would be nice if commits in PR does not add code in one commit, just to remove it again at later commit.

Also, I think there should be doc followup, as I don't see any docs describing this feature.

nordic-krch commented:
@tejlmand I will add doc in separate PR
